### ISSUE TITLE
The implementation of BPR loss is different from that stated in the paper

### ISSUE THREAD

**Issue by Dinxin** [gusye1234/LightGCN-PyTorch#4]
The code implementation of BPR loss:
```
def bpr_loss(self, users, pos, neg):
        (users_emb, pos_emb, neg_emb, 
        userEmb0,  posEmb0, negEmb0) = self.getEmbedding(users.long(), pos.long(), neg.long())
        reg_loss = (1/2)*(userEmb0.norm(2).pow(2) + 
                         posEmb0.norm(2).pow(2)  +
                         negEmb0.norm(2).pow(2))/float(len(users))
        pos_scores = torch.mul(users_emb, pos_emb)
        pos_scores = torch.sum(pos_scores, dim=1)
        neg_scores = torch.mul(users_emb, neg_emb)
        neg_scores = torch.sum(neg_scores, dim=1)
    
        loss = torch.mean(torch.nn.functional.softplus(neg_scores - pos_scores))
            
        return loss, reg_loss
```
The formula stated in the paper:
![image](https://user-images.githubusercontent.com/19572535/92217306-1f0d6f00-eeca-11ea-98bd-69e9522d7c46.png)

**Comment by gusye1234** [comment]
Do you have a question about this line
```python
reg_loss = (1/2)*(userEmb0.norm(2).pow(2) + 
                         posEmb0.norm(2).pow(2)  +
                         negEmb0.norm(2).pow(2))/float(len(users))
```
with a 1/2 in front of the regularization term?
I think that's just a non-important difference since we can always absorb the 1/2 into lambda.
